File: AOJ/AOJ vol.16/aoj1627_WA.cpp
#include <bits/stdc++.h>
using namespace std;
#define REP(i,b,e) for(int i=(b);i<(e);++i)
#define rep(i,e) for(int i=0;i<(e);++i)

int main() {
  int n;
  while(cin>>n){
    if(n==0) return 0;

    int draw = (n-1)/2;

    vector<vector<int>> match(n, vector<int>(n, 0));
    vector<int> wins(n, 0);

    int m; cin >> m;
    rep(i, m){
      int x, y;
      cin >> x >> y;
      x--;
      y--;
      match[x][y] = 1;
      match[y][x] = -1;
      wins[x] += 1;
    }

    int ans = 0;
    int rm = n*(n-1)/2-m;
    rep(i, 1<<rm){
      int cur = 0;
      bool good = true;
      vector<int> wint = wins;
      rep(j, n-1){
        if(good==false || cur>=rm) break;
        REP(k, j+1, n){
          if(wint[j]>draw || wint[k]>draw){
            good = false;
            break;
          }
          if(cur>=rm) break;
          if(match[j][k]==0){
            if((i&(1<<cur))!=0){
              wint[j]++;
              if(wint[j] > draw){
                good = false;
                break;
              }
            }
            else{
              wint[k]++;
              if(wint[k] > draw){
                good = false;
                break;
              }
            }
            cur++;
          }
        }
      }
      if(good) ans+=1;
    }
    cout<<ans<<endl;
  }

  return 0;
}

File: AOJ/AOJ vol.16/aoj1627_WA.py
import copy
from collections import defaultdict
while True:
  n = int(input())
  if n==0:
    break

  draw = (n-1)//2

  match = [[0 for i in range(n+1)] for j in range(n+1)]
  win = defaultdict(int)
  m = int(input())
  for i in range(m):
    x, y = list(map(int, input().split()))
    match[x][y] = 1
    match[y][x] = -1
    win[x] += 1

  ans = 0
  rm = (n*(n-1)//2-m)
  for i in range(1<<rm):
    cur = 0
    good = True
    wint = copy.deepcopy(win)
    for j in range(1, n):
      if not good or cur>=rm:
        break
      for k in range(j+1, n+1):
        if wint[j] > draw or wint[k] > draw:
          good = False
          break
        if cur>=rm:
          break
        if match[j][k]==0:
          if i&(1<<cur)!=0:
            wint[j] += 1
            if wint[j] > draw:
              good = False
              break
          else:
            wint[k] += 1
            if wint[k] > draw:
              good = False
              break
          cur += 1
    if good:
      ans += 1

  print(ans)